activate

```diff
@@ -1,95 +1,218 @@
+            except Exception as e:
+                print("Unable to get / image name!", e, file=sys.stderr)
+                return image
+
+            print(" - Image root:", image_root)
+
+            if image_root != "":
+                return image_root
+            else:
+                image
+
+        return None
+
+    def _get_release(self):
+        release = None
+
+        try:
+            release_data = None
+            with open('/etc/os-release') as release_file:
+                release_data = release_file.read()
+
+            release_data = re.split("\n", release_data, flags=re.M)
+            for release_item in release_data:
+                release_item = release_item.strip()
+
+                key, value = release_item.split('=', 1)
+                if key == 'VERSION':
+                    print(' - Found version key')
+                    if value.startswith('"') and value.endswith('"'):
+                            value = value[1:-1]
+
+                    release = value
+                    print(' - Release:', release)
+                    return release
+
+        except Exception as e:
+            print("Unable to get release name!", e, file=sys.stderr)
+
+        return release.strip()
+
+    def _get_serial(self):
+        serial = None
+
+        try:
+            if os.path.isdir('/sys/class/dmi/id'):
+                with open('/sys/class/dmi/id/product_serial') as ssn_file:
+                    serial = ssn_file.read().strip()
+            else:
+                with open('/sys/class/endless_mfgdata/entries/SSN') as ssn_file:
+                    serial = ssn_file.read().strip()
+        except Exception as e:
+            print("Unable to get serial! Check that you're running this as root!", e, file=sys.stderr)
+
+        print(" - Found serial:", serial)
+
+        return serial.strip()
+
+    def _get_is_live(self):
+        is_live = False
+
+        command_line = None
+        with open('/proc/cmdline', 'r') as command_line_file:
+            command_line=command_line_file.read()
+
+        print(" - Cmdline:", command_line.strip())
+
+        if re.search(r'\bendless.live\b', command_line):
+            print(" - Found is_live marker. Marking as live image.")
+            is_live = True
+
+        return is_live
+
+    def _send_to_server(self, activation_data):
+        success = False
+        print("Sending data...")
+
+        try:
+            data = activation_data.encode('utf-8')
+
+            req = urllib.request.Request(self.ACTIVATION_ENDPOINT, data = data, method = 'PUT')
+            req.add_header('Content-Type', 'application/json')
+
+            response_data = None
+            with urllib.request.urlopen(req) as conn:
+                response_data = conn.read().decode('utf-8')
+
+            print('Response:', response_data)
+
+            response = json.loads(response_data)
```

Comment:
I was about to say maybe we should check the mime type and log something more clearly like "got a reply I didn't like" because of captive portals, but we'd hope we actually get a certificate error in that case.